From 38402fa6c5bfcf9a4a9b143dbc710f4b033ffdb2 Mon Sep 17 00:00:00 2001
From: Sriram Dash <quic_sriramd@quicinc.com>
Date: Thu, 27 Feb 2025 20:49:10 +0530
Subject: [PATCH 3/5] PENDING: usb: dwc3: qcom: Add support for sa8255p for
 qcom usb

QCOM SoC sa8255p allow multiple resources to be managed
by firmware. On this platform, tasks related to clocks, regulators,
resets, TLMM and interconnects can be delegated to the firmware,
while the remaining responsibilities are handled by Linux.

The power domain, managed through the PM runtime framework,
is utilized for activation/deactivation of the resources.

Power domain "usb_core" controls the GDSC, resets,
and TLMM(gpio) whereas power domain "usb_transfer"
controls clocks and interconnects.

On this platform, the two disjoint sets of resources are modeled against
the power domain suppliers which can be turned on/off to achieve modeled
device states. Two modeled low power states d1 and d3 are supported.

transition: d0 <--> d3  :: both core and transfer suppliers
transition: d0 <--> d1  :: only transfer supplier

Change-Id: Iaa236081cd76d263279d23115534bd2b81ec70ba
Signed-off-by: Sriram Dash <quic_sriramd@quicinc.com>
Signed-off-by: Shazad Hussain <quic_shazhuss@quicinc.com>
---
 drivers/usb/dwc3/dwc3-qcom.c | 129 ++++++++++++++++++++++++++++++++++-
 1 file changed, 126 insertions(+), 3 deletions(-)

diff --git a/drivers/usb/dwc3/dwc3-qcom.c b/drivers/usb/dwc3/dwc3-qcom.c
index c5ceddae6f5d9..952df33e28fb6 100644
--- a/drivers/usb/dwc3/dwc3-qcom.c
+++ b/drivers/usb/dwc3/dwc3-qcom.c
@@ -16,6 +16,7 @@
 #include <linux/interconnect.h>
 #include <linux/of_platform.h>
 #include <linux/platform_device.h>
+#include <linux/pm_domain.h>
 #include <linux/phy/phy.h>
 #include <linux/usb/of.h>
 #include <linux/reset.h>
@@ -63,6 +64,10 @@ struct dwc3_acpi_pdata {
 	bool			is_urs;
 };
 
+struct dwc3_qcom_priv_data {
+	bool	fw_managed;
+};
+
 struct dwc3_qcom {
 	struct device		*dev;
 	void __iomem		*qscratch_base;
@@ -95,8 +100,83 @@ struct dwc3_qcom {
 	bool			ignore_pipe_clk;
 	enum usb_role		current_role;
 	struct notifier_block	xhci_nb;
+
+	struct dev_pm_domain_list	*pd_list;
+};
+
+static const struct dwc3_qcom_priv_data sa8255p_dwc3_qcom_priv_data = {
+	.fw_managed	= true,
 };
 
+static void dwc3_qcom_domain_detach(struct dwc3_qcom *qcom)
+{
+	dev_pm_domain_detach_list(qcom->pd_list);
+}
+
+static int dwc3_qcom_domain_attach(struct dwc3_qcom *qcom)
+{
+	struct dev_pm_domain_attach_data pd_data = {
+		.pd_flags	= PD_FLAG_NO_DEV_LINK,
+		.pd_names	= (const char*[]) { "usb_transfer", "usb_core" },
+		.num_pd_names	= 2,
+	};
+	struct device *dev = qcom->dev;
+	int ret = 0;
+
+	ret = dev_pm_domain_attach_list(dev, &pd_data, &qcom->pd_list);
+	if (ret < 0) {
+		dev_err(dev, "domain attach failed %d)\n", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+/* d3_to_d0 transition by turning on all the suppliers */
+static int dwc3_qcom_d3_to_d0(struct dwc3_qcom *qcom)
+{
+	int ret = 0;
+
+	if (!qcom->pd_list)
+		return 0;
+
+	ret = pm_runtime_resume_and_get(qcom->pd_list->pd_devs[1]);
+
+	ret = pm_runtime_resume_and_get(qcom->pd_list->pd_devs[0]);
+	if (ret)
+		return ret;
+
+	return ret;
+}
+
+/* d0_to_d3 transition by turning off all the suppliers */
+static void dwc3_qcom_d0_to_d3(struct dwc3_qcom *qcom)
+{
+	if (!qcom->pd_list)
+		return;
+
+	pm_runtime_put_sync(qcom->pd_list->pd_devs[0]);
+	pm_runtime_put_sync(qcom->pd_list->pd_devs[1]);
+}
+
+/* d1_to_d0 transition by turning on the 'tranfer' supplier */
+static int dwc3_qcom_d1_to_d0(struct dwc3_qcom *qcom)
+{
+	if (!qcom->pd_list)
+		return 0;
+
+	return pm_runtime_resume_and_get(qcom->pd_list->pd_devs[0]);
+}
+
+/* d0_to_d1 transition by turning off the 'tranfer' supplier */
+static void dwc3_qcom_d0_to_d1(struct dwc3_qcom *qcom)
+{
+	if (!qcom->pd_list)
+		return;
+
+	pm_runtime_put_sync(qcom->pd_list->pd_devs[0]);
+}
+
 static inline void dwc3_qcom_setbits(void __iomem *base, u32 offset, u32 val)
 {
 	u32 reg;
@@ -476,6 +556,10 @@ static int dwc3_qcom_suspend(struct dwc3_qcom *qcom, bool wakeup)
 	val = readl(qcom->qscratch_base + PWR_EVNT_IRQ_STAT_REG);
 	if (!(val & PWR_EVNT_LPM_IN_L2_MASK))
 		dev_err(qcom->dev, "HS-PHY not in L2\n");
+	if (wakeup)
+		dwc3_qcom_d0_to_d1(qcom);
+	else
+		dwc3_qcom_d0_to_d3(qcom);
 
 	for (i = qcom->num_clocks - 1; i >= 0; i--)
 		clk_disable_unprepare(qcom->clks[i]);
@@ -519,6 +603,16 @@ static int dwc3_qcom_resume(struct dwc3_qcom *qcom, bool wakeup)
 	if (dwc3_qcom_is_host(qcom) && wakeup)
 		dwc3_qcom_disable_interrupts(qcom);
 
+	if (wakeup)
+		ret = dwc3_qcom_d1_to_d0(qcom);
+	else
+		ret = dwc3_qcom_d3_to_d0(qcom);
+
+	if (ret < 0) {
+		dev_err(qcom->dev, "Failed to transition to d0 state\n");
+		return ret;
+	}
+
 	for (i = 0; i < qcom->num_clocks; i++) {
 		ret = clk_prepare_enable(qcom->clks[i]);
 		if (ret < 0) {
@@ -728,6 +822,7 @@ static int dwc3_qcom_clk_init(struct dwc3_qcom *qcom, int count)
 
 static const struct property_entry dwc3_qcom_props[] = {
 	PROPERTY_ENTRY_BOOL("snps,allow-role-switch-userspace-control"),
+	PROPERTY_ENTRY_BOOL("xhci-reset-on-resume"),
 	{ },
 };
 
@@ -977,6 +1072,7 @@ static int dwc3_qcom_probe(struct platform_device *pdev)
 	struct device_node	*np = pdev->dev.of_node;
 	struct device		*dev = &pdev->dev;
 	struct dwc3_qcom	*qcom;
+	const struct dwc3_qcom_priv_data	*priv_data;
 	struct resource		*res, *parent_res = NULL;
 	struct resource		local_res;
 	int			ret, i;
@@ -988,6 +1084,8 @@ static int dwc3_qcom_probe(struct platform_device *pdev)
 	if (!qcom)
 		return -ENOMEM;
 
+	priv_data = of_device_get_match_data(dev);
+
 	legacy_binding = dwc3_qcom_has_separate_dwc3_of_node(dev);
 
 	if (!legacy_binding)
@@ -1002,6 +1100,20 @@ static int dwc3_qcom_probe(struct platform_device *pdev)
 	if (ret)
 		return ret;
 
+	if (priv_data && priv_data->fw_managed) {
+		ret = dwc3_qcom_domain_attach(qcom);
+		if (ret) {
+			dev_err(dev, "Failed to attach domains. Bail out\n");
+			return ret;
+		}
+	}
+
+	ret = dwc3_qcom_d3_to_d0(qcom);
+	if (ret < 0) {
+		dev_err(qcom->dev, "Failed to transition to d0 state\n");
+		goto domain_detach;
+	}
+
 	if (has_acpi_companion(dev)) {
 		qcom->acpi_pdata = acpi_device_get_match_data(dev);
 		if (!qcom->acpi_pdata) {
@@ -1025,14 +1137,13 @@ static int dwc3_qcom_probe(struct platform_device *pdev)
 	if (legacy_binding) {
 		qcom->resets = devm_reset_control_array_get_optional_exclusive(dev);
 		if (IS_ERR(qcom->resets)) {
-			return dev_err_probe(&pdev->dev, PTR_ERR(qcom->resets),
-					     "failed to get resets\n");
+			goto resources_off;
 		}
 
 		ret = reset_control_assert(qcom->resets);
 		if (ret) {
 			dev_err(&pdev->dev, "failed to assert resets, err=%d\n", ret);
-			return ret;
+			goto resources_off;
 		}
 
 		usleep_range(10, 1000);
@@ -1157,6 +1268,11 @@ static int dwc3_qcom_probe(struct platform_device *pdev)
 reset_assert:
 	reset_control_assert(qcom->resets);
 
+resources_off:
+	dwc3_qcom_d0_to_d3(qcom);
+
+domain_detach:
+	dwc3_qcom_domain_detach(qcom);
 	return ret;
 }
 
@@ -1181,6 +1297,9 @@ static void dwc3_qcom_remove(struct platform_device *pdev)
 	else
 		platform_device_put(pdev);
 
+	dwc3_qcom_d0_to_d3(qcom);
+	dwc3_qcom_domain_detach(qcom);
+
 	for (i = qcom->num_clocks - 1; i >= 0; i--) {
 		clk_disable_unprepare(qcom->clks[i]);
 		clk_put(qcom->clks[i]);
@@ -1308,6 +1427,10 @@ static const struct dev_pm_ops dwc3_qcom_dev_pm_ops = {
 };
 
 static const struct of_device_id dwc3_qcom_of_match[] = {
+	{
+		.compatible	= "qcom,sa8255p-dwc3",
+		.data		= &sa8255p_dwc3_qcom_priv_data,
+	},
 	{ .compatible = "qcom,dwc3" },
 	{ }
 };
-- 
2.17.1

